### DISCUSSION TITLE
Properties passed into Business Rule Constructor seem to automatically be Dependent/Affected Properties

### DISCUSSION THREAD

**Discussion by mtavares628** [MarimerLLC/csla#3415] [Questions]
I've noticed this issue which ends up causing the same rules on my object to be run over and over.  Basically if I pass properties (besides the PrimaryProperty) to the constructor of my rule (Rule A), and that passed in property is used as a PrimaryProperty in some other rule (Rule B), then Rule A gets called again whenever Rule B has been called. I would expect this to happen if I added the property to the AffectedProperties list for the rule, or added a Dependency rule between the 2 properties, but I'm not doing that.  

Here is a sample that recreates the behavior.  If you set a breakpoint on ExecuteAsync of the IDRule below, then it will get called twice when CheckRules is called.

```c#
[Serializable]
    public class Root : BusinessBase<Root>
    {
        public static readonly PropertyInfo<string> NameProperty = RegisterProperty<string>(nameof(Name));
        public string Name
        {
            get => GetProperty(NameProperty);
            set => SetProperty(NameProperty, value);
        }

        public static readonly PropertyInfo<Csla.SmartDate> ReferenceDateProperty = RegisterProperty<Csla.SmartDate>(nameof(ReferenceDate));

        public string ReferenceDate
        {
            get => GetPropertyConvert<Csla.SmartDate, string>(ReferenceDateProperty);
            set => SetPropertyConvert<Csla.SmartDate, string>(ReferenceDateProperty, value);
        }

        public static readonly PropertyInfo<int> IDProperty = RegisterProperty<int>(nameof(ID));

        public int ID
        {
            get => GetProperty(IDProperty);
            set => SetProperty(IDProperty, value);
        }
        


        protected override void AddBusinessRules()
        {
            base.AddBusinessRules();
            BusinessRules.AddRule(new ReferenceDateRule(ReferenceDateProperty));
            BusinessRules.AddRule(new IDRule(IDProperty, ReferenceDateProperty));

            //BusinessRules.AddRule(new MyRule { PrimaryProperty = NameProperty });
        }

        [Fetch]
        private void Fetch() 
        {
            using (BypassPropertyChecks)
            {
                Name = "Test";
                ReferenceDate = DateTime.Today.ToShortDateString();
                ID = 1;
            }
            BusinessRules.CheckRules();
        }
    }

public class ReferenceDateRule : Csla.Rules.BusinessRuleAsync
    {
        //IPropertyInfo IDProperty;

        public ReferenceDateRule(IPropertyInfo primaryProp) : base(primaryProp)
        {
            InputProperties = new List<IPropertyInfo>() { PrimaryProperty };

        }

        protected override async Task ExecuteAsync(IRuleContext context)
        {
            // breakpoint vvv
            context.AddSuccessResult(false);
        }
    }

    public class IDRule : Csla.Rules.BusinessRuleAsync
    {
        IPropertyInfo ReferenceDateProperty;

        public IDRule(IPropertyInfo primaryProp, IPropertyInfo referenceDateProp) : base(primaryProp)
        {
            ReferenceDateProperty = referenceDateProp;

            InputProperties = new List<IPropertyInfo>() { PrimaryProperty, ReferenceDateProperty };

        }

        protected override async Task ExecuteAsync(IRuleContext context)
        {
            // breakpoint vvv
            context.AddSuccessResult(false);
        }
    }

```

Is this by design?  And if so, then how can I create rules with multiple properties, and prevent the rule from being called multiple times if the secondary (non-PrimaryProperty) has it's own set of rules?  I can't use the Target object either, since my rules need to be Async for Blazor.

**Comment by Chicagoan2016** [comment]
I know CommonBusinessRule, which inherites from PropertyRule, has 'CanRunAsAffectedProperty'
You could implement something similar in ASync version.

**Comment by rockfordlhotka** [comment]
Using this run code:

```c#
using Csla;
using Csla.Configuration;
using Microsoft.Extensions.DependencyInjection;

var services = new ServiceCollection();
services.AddCsla();
var provider  = services.BuildServiceProvider();
var dp = provider.GetRequiredService<IDataPortal<Root>>();

var root = await dp.FetchAsync();
root.ID = 123;
root.ReferenceDate = DateTime.Today.ToString();
```

I'm seeing the `IDRule` run three times.

Twice due to the `CheckRules` call during fetch and once when the ID property is set.

It doesn't get triggered when the reference date is set.

Why `CheckRules` is running it twice isn't clear to me.

**Reply by mtavares628** [reply]
@rockfordlhotka - In my troubleshooting, it has something to do with the combination of having the ReferenceDateProperty as a secondary property on the one rule AND ReferenceDateProperty being the Primary on another rule.  If you comment out the ReferenceDateRule in AddBusinessRules, then CheckRules during fetch won't run the IDRule twice anymore.  My issue is that I have a number of rules with multiple properties like this, and some rules end up running 20 times on fetch.

**Reply by mtavares628** [reply]
@rockfordlhotka - When I run your code, I actually see the following:

 IDRule running 4 times (twice during fetch, once for the ID change and once for the date change).
ReferenceDateRule runs twice (once during fetch, once for the date change).

In stepping through the code, the culprit seems to be the following code in the BusinessRules.CheckRulesForProperty function:

```c#
private List<string> CheckRulesForProperty(Csla.Core.IPropertyInfo property, bool cascade, RuleContextModes executionContext)
    {
      var rules = from r in TypeRules.Rules
                  where ReferenceEquals(r.PrimaryProperty, property)
                    && CanRunRule(ApplicationContext, r, executionContext)
                  orderby r.Priority
                  select r;

      BrokenRules.ClearRules(property);
      var firstResult = RunRules(rules, cascade, executionContext);
      if (CascadeOnDirtyProperties)
          cascade = cascade || firstResult.DirtyProperties.Any();
      if (cascade)
      {
        // get properties affected by all rules
        var propertiesToRun = new List<Csla.Core.IPropertyInfo>();
        foreach (var item in rules)
          if (!item.IsAsync)
          {
            foreach (var p in item.AffectedProperties)
              if (!ReferenceEquals(property, p))
                propertiesToRun.Add(p);
          }

        // add PrimaryProperty where property is in InputProperties
        var input = from r in TypeRules.Rules
                    where !ReferenceEquals(r.PrimaryProperty, property)
                          && r.PrimaryProperty != null
                          && r.InputProperties != null
                          && r.InputProperties.Contains(property)
                    select r.PrimaryProperty;

        foreach (var p in input)
        {
            if (!ReferenceEquals(property, p))
                propertiesToRun.Add(p);
        }
        // run rules for affected properties
        foreach (var item in propertiesToRun.Distinct())
        {
          var doCascade = false; 
          if (CascadeOnDirtyProperties)
            doCascade = firstResult.DirtyProperties.Any(p => p == item.Name);
          firstResult.AffectedProperties.AddRange(CheckRulesForProperty(item, doCascade,
                                                               executionContext | RuleContextModes.AsAffectedPoperty));
        }
      }

      // always make sure to add PrimaryProperty
      firstResult.AffectedProperties.Add(property.Name);
      return firstResult.AffectedProperties.Distinct().ToList();
    }
```
In particular this section:

```c#
// add PrimaryProperty where property is in InputProperties
        var input = from r in TypeRules.Rules
                    where !ReferenceEquals(r.PrimaryProperty, property)
                          && r.PrimaryProperty != null
                          && r.InputProperties != null
                          && r.InputProperties.Contains(property)
                    select r.PrimaryProperty;

        foreach (var p in input)
        {
            if (!ReferenceEquals(property, p))
                propertiesToRun.Add(p);
        }
        // run rules for affected properties
        foreach (var item in propertiesToRun.Distinct())
        {
          var doCascade = false; 
          if (CascadeOnDirtyProperties)
            doCascade = firstResult.DirtyProperties.Any(p => p == item.Name);
          firstResult.AffectedProperties.AddRange(CheckRulesForProperty(item, doCascade,
                                                               executionContext | RuleContextModes.AsAffectedPoperty));
        }
```
It looks like it is grabbing all other properties from all rules where the passed in property (ReferenceDateProperty in this case) has been added to the InputProperties list.  I'm not sure why we would need to run all other rules there. Should it be looking to make sure that each of those InputProperties is also in the AffectedProperties list?

I'm also not sure why the AffectedProperties aren't being added if the rule is Async:

```c#
if (!item.IsAsync)
          {
            foreach (var p in item.AffectedProperties)
              if (!ReferenceEquals(property, p))
                propertiesToRun.Add(p);
          }
```
And one other thing I've noticed is that in this function we are always forcing it to cascade the dependent properties (true is being passed to the cascade parameter of CheckRulesForProperty):

```c#
private  List<string> CheckRules(Csla.Core.IPropertyInfo property, RuleContextModes executionContext)
    {
      if (property == null)
        throw new ArgumentNullException(nameof(property));

      if (_suppressRuleChecking)
        return new List<string>();

      var oldRR = RunningRules;
      RunningRules = true;

      var affectedProperties = new List<string>();
      affectedProperties.AddRange(CheckRulesForProperty(property, true, executionContext));

      RunningRules = oldRR;
      if (!RunningRules && !RunningAsyncRules)
        _target.AllRulesComplete();
      return affectedProperties.Distinct().ToList();
    }

```

This function is being called from the overall CheckRules method that is running for all properties anyway:

```c#
public List<string> CheckRules()
    {
      if (_suppressRuleChecking)
        return new List<string>();

      RunningRules = true;
      var affectedProperties = CheckObjectRules(RuleContextModes.CheckRules, false);
      var properties = TypeRules.Rules.Where(p => p.PrimaryProperty != null)
                                          .Select(p => p.PrimaryProperty)
                                          .Distinct();
      foreach (var property in properties)
        affectedProperties.AddRange(CheckRules(property, RuleContextModes.CheckRules));
      RunningRules = false;
      if (!RunningRules && !RunningAsyncRules)
        _target.AllRulesComplete();
      return affectedProperties.Distinct().ToList();
    }
```
Shouldn't the cascade be turned off at that point, since we are already guaranteeing that all of the rules are being run?

I may be missing reasons for all of these, hence why I'm asking.

**Reply by mtavares628** [reply]
Upon further digging, it looks like this functionality was actually introduced all the way back in v4.2.  Hence, it isn't in the CSLA 4 book and the only info I could find on it is in Jonny Bee's blog post: [https://jonnybekkum.wordpress.com/2011/08/29/csla-4-2-rules-update/](https://jonnybekkum.wordpress.com/2011/08/29/csla-4-2-rules-update/)

It looks like it was done for convenience of creating the dependencies. I never noticed this functionality before because a lot of my rules that utilized multiple properties used the Target object instead, and they were all synchronous, since everything ran server side.  

However, with the advent of Blazor and all Data Portal calls being required to run asynchronously, I needed to alter my rules to stop using the Target object, and instead pass them in as InputProperties.  This has triggered all of these repeated rule checks, which is slowing my app down, since it's running on the client and keeps having to run a remote data portal on the server.

I would also argue that there are a number of situations where you wouldn't want to make all other Input Properties affected properties.  For example, let's say we have 3 properties (A, B and C).  Each of these have there own sets of rules where they are the primary property.  But there is one rule where property C is dependent on both A and B.  So C is the Primary, and A and B are passed in as InputProperties to check their values and do something with C.  However, A and B have no actual bearing on one another.  But with the All Input Properties logic, they all become dependent on one another.  So when I make a change to A, it calls A's Rules (normal), C's Rules (expected) and B's Rules (which are not needed because A doesn't affect B).  

Is there something we can do to change this functionality, since Async Rules are more prevalent and performance can be hindered by recalling all of these rules multiple times across the wire?  Users would still have both the Dependency Rule and the Affected Properties to handle those dependencies, but it would allow an option to not make something dependent, since Target can't be used in Async rules.

I understand this would be a breaking change, but I think it's a necessary one, unless you have a different workaround for this.  Plus this current functionality, is not very well documented outside of that blog post.

**Reply by rockfordlhotka** [reply]
That's some great research @mtavares628 !!

Any thoughts @jonnybee ?

I agree that this should be something you can control. Or perhaps we need a way to optimize rules that run via `CheckRules`/`CheckRulesAsync` to avoid/minimize repeats.

No doubt that this is a breaking change - yet now is the time, because CSLA 8 is my current work target, and that'll incluce breaking changes, and should come out shortly after .NET 8 in a couple months.

**Reply by mtavares628** [reply]
@rockfordlhotka and @jonnybee, has any thought been given to this situation so that properties added to a rule aren't automatically set as a dependent/affected property?  I'd like to see this incorporated into CSLA 8, since the performance with running async rules many times is not great.  Thanks.

**Reply by mtavares628** [reply]
So in looking at this some more, I have a question and some possible solutions.

Could we have things run the other way in which the Input Properties act as a trigger for that particular rule, instead of triggering the rules for the other Input Properties?  To me this would cut down the same rule being triggered multiple times due to the affected properties, but still only trigger for the properties that truly affect it.

Here are some of my proposed options:

- Add a boolean flag to the rule named something like InputPropertiesRunAsAffectedProperties.  You can default it to true for backwards compatibility, but if false, it doesn't run the routine to add all of the input properties as affected properties.
- Add a TriggeredProperties collection to the rule. Any of the properties added to it would then trigger that rule in addition to the Primary Property.

So the CheckRulesForProperty function would look something like this:

```c#
                private List<string> CheckRulesForProperty(Csla.Core.IPropertyInfo property, bool cascade, RuleContextModes executionContext)
        {
            var rules = from r in TypeRules.Rules
                        where ReferenceEquals(r.PrimaryProperty, property)
                          && CanRunRule(ApplicationContext, r, executionContext)
                        orderby r.Priority
                        select r;

            BrokenRules.ClearRules(property);
            var firstResult = RunRules(rules, cascade, executionContext);

            // Run any rules that are triggered from this property
            var triggeredRules = from r in TypeRules.Rules
                                 where !ReferenceEquals(r.PrimaryProperty, property)
                                       && r.PrimaryProperty != null
                                       && r.TriggeredProperties != null
                                       && r.TriggeredProperties.Contains(property)
                                       && CanRunRule(ApplicationContext, r, executionContext)
                                 select r;

            // Not sure if we should cascade here or not, or whether we need to use the rules result for anything
            RunRules(triggeredRules, false, executionContext);

            if (CascadeOnDirtyProperties)
                cascade = cascade || firstResult.DirtyProperties.Any();
            if (cascade)
            {
                // get properties affected by all rules
                var propertiesToRun = new List<Csla.Core.IPropertyInfo>();
                foreach (var item in rules)
                    if (!item.IsAsync)
                    {
                        foreach (var p in item.AffectedProperties)
                            if (!ReferenceEquals(property, p))
                                propertiesToRun.Add(p);
                    }

                // add PrimaryProperty where property is in InputProperties
                if (InputPropertiesRunAsAffectedProperties)
                {
                    var input = from r in TypeRules.Rules
                                where !ReferenceEquals(r.PrimaryProperty, property)
                                      && r.PrimaryProperty != null
                                      && r.InputProperties != null
                                      && r.InputProperties.Contains(property)
                                select r.PrimaryProperty;

                    foreach (var p in input)
                    {
                        if (!ReferenceEquals(property, p))
                            propertiesToRun.Add(p);
                    }
                }
                // run rules for affected properties
                foreach (var item in propertiesToRun.Distinct())
                {
                    var doCascade = false;
                    if (CascadeOnDirtyProperties)
                        doCascade = firstResult.DirtyProperties.Any(p => p == item.Name);
                    firstResult.AffectedProperties.AddRange(CheckRulesForProperty(item, doCascade,
                                                                         executionContext | RuleContextModes.AsAffectedPoperty));
                }
            }

            // always make sure to add PrimaryProperty
            firstResult.AffectedProperties.Add(property.Name);
            return firstResult.AffectedProperties.Distinct().ToList();
        }
```

I'm not 100% sure whether you would want to cascade on the Triggered Property Rules or not. My guess would be not, since you are only running those rules that the changed property has affected. Maybe I'm way off base here, but this would cut down on a lot of the rule re-running if I only triggered the rules that pertain to that specific property rather than running every rule for any affected property.  @rockfordlhotka I welcome your thoughts, and why this may not be a good idea.

**Reply by ossendorf-at-hoelscher** [reply]
Anything new here?
We are having issues with having AsyncRules executed concurrently for a non-dependent property from which we only need the value.

**Reply by ossendorf-at-hoelscher** [reply]
Furthermore if you have an async rule and the properties A, B, C. Where A is the primary property, B & C are input properties. And also C is an affected property.

When you do the following:
```csharp
// f.B is any value not set here
f.A = "Abcd";
f.C = "cdef";
```
The async rule is executed concurrently and as far as I can see. Always the first rule execution (triggerd from the assignment of A) wins the  race even though I set C explicitly (which triggers the rule again).

**Reply by mtavares628** [reply]
Issue #3616 addressed this.  @rockfordlhotka  added a check to the CheckRulesForProperty code that only runs the rule on the Input Property if that property was actually changed (in the dirty properties collection).  This definitely cut down on the multiple running of rules for me.  The fix is in CSLA 8 prereleases.  

I'm still considering putting in a pull request for my TriggeredProperties concept, because I think it allows a user full control over when that rule will get called right within the rule itself, but I haven't gone down that road of trying that concept out yet.
